src/vs/workbench/api/common/extHostNotebookDocument.ts

```diff
@@ -114,12 +114,16 @@ export class ExtHostCell {
 
 	setOutputItems(outputId: string, append: boolean, newOutputItems: extHostProtocol.NotebookOutputItemDto[]) {
 		const newItems = newOutputItems.map(extHostTypeConverters.NotebookCellOutputItem.to);
-		const output = this._outputs.find(op => op.id === outputId);
-		if (output) {
+
+		// Use an index so we can overwrite internal properties (otherwise readonly)
+		const index = this._outputs.findIndex(op => op.id === outputId);
+		if (index >= 0) {
+			const output = deepClone(this._outputs[index]);
 			if (!append) {
```

Comment:
Not very familiar with this code so trying to understand why the deep clone is correct

Can you share more details about what is happening on the python side that causes the error?  My guess is that the existing code mutates the `NotebookCellOutput.items`, and some code is perhaps taking a reference to `NotebookCellOutput.items` and expecting it not to change behind the scenes? Is that correct?

By cloning the entire object, doesn't that also mean that after every call to `setOutputItems`, the `NotebookCellOutput` we expose in  the api is now a different object (even if it has the same id)? What happens if an extension is holding onto an old `NotebookCellOutput` at that point?

Reply:
I believe it's failing because of the freeze that happens when sending the event. An alternative would be to not freeze at all in when sending the event.

This call here crashes:
```
output.items.length = 0
```
With an error stating a readonly object cannot be modified (likely because it was frozen).

Reply:
What about the second point? By switching from mutating the original array to replacing one output with another, is there a risk of having multiple versions of the same output item that extensions are hanging on to?